src/core/model_event_queue.jl

```diff
@@ -0,0 +1,278 @@
+can be evolved in time.
+(you can always use this function regardless of the default event scheduling behavior)
+
+## Keywords
+- `container, properties, rng, warn`: same as in [`StandardABM`](@ref).
+- `autogenerate_on_add::Bool = true`: whether to automatically generate a new event for
+  an agent when the agent is added to the model.
+- `autogenerate_after_action::Bool = true`: whether to automatically generate a new
+  event for an agent after an event affected said agent has been triggered.
+"""
+function EventQueueABM(
+        ::Type{A}, events::E,
+        space::S = nothing;
+        container::Type = Dict{Int},
+        properties::P = nothing,
+        rng::R = Random.default_rng(),
+        warn = true,
+        autogenerate_on_add = true,
+        autogenerate_after_action = true,
+    ) where {A<:AbstractAgent,S<:SpaceType,E,P,R<:AbstractRNG}
+    agent_validator(A, space, warn)
+    C = construct_agent_container(container, A)
+    agents = C()
+    I = events isa Tuple ? Int : keytype(events)
+    # The queue stores references to events;
+    # the reference is two integers; one is the agent ID
+    # and the other is the index of the event in `events`
+    queue = PriorityQueue{Tuple{I, Int}, Float64}()
+    propensities = zeros(length(events))
+    return EventQueueABM{S,A,C,P,E,R,I}(
+        Ref(0.0), agents, space, properties, rng,
+        Ref(0), events, propensities, queue,
+        autogenerate_on_add,
+        autogenerate_after_action,
+    )
+end
+
+# standard accessors
+"""
+    abmqueue(model::EventQueueABM)
+
+Return the queue of scheduled events in the `model`.
+The queue maps two integers (agent id, event index) to
+the time the event will occur, in absolute time.
+"""
+abmqueue(model::EventQueueABM) = getfield(model, :event_queue)
+
+"""
+    abmevents(model::EventQueueABM)
+
+Return all possible events stored in the model.
+"""
+abmevents(model::EventQueueABM) = getfield(model, :events)
+
+"""
+    abmtime(model::AgentBasedModel)
+
+Return the current time of the model.
+All models are initialized at time 0.
+"""
+abmtime(model::EventQueueABM) = getfield(model, :time)[]
+
+containertype(::EventQueueABM{S,A,C}) where {S,A,C} = C
+agenttype(::EventQueueABM{S,A}) where {S,A} = A
+
+###########################################################################################
+# %% Adding events to the queue
+###########################################################################################
+# This function ensures that once an agent is added into the model,
+# an event is created and added for it. It is called internally
+# by `add_agent_to_model!`.
+function extra_actions_after_add!(agent, model::EventQueueABM)
+    if getfield(model, :autogenerate_on_add)
+        generate_event_in_queue!(agent, model)
+    end
+end
+
+# This is the main function that is called in `step!`
+# after the other main function `process_event!` is
+# done
+"""
+    generate_event_in_queue!(agent, model)
```

Comment:
I propose to dispatch this to `add_event!`

```suggestion
    add_event!(agent, model)
```